assets/src/edit-story/app/media/media3p/useFetchMediaEffect.js

```diff
@@ -53,41 +57,66 @@ export default function useFetchMediaEffect({
     actions: { listMedia, listCategoryMedia },
   } = useMedia3pApi();
 
+  const previousProps = usePrevious({
+    pageToken,
+    searchTerm,
+    selectedCategoryId,
+  });
+
   const { showSnackbar } = useSnackbar();
 
   useEffect(() => {
     async function fetch() {
       fetchMediaStart({ provider, pageToken });
       try {
-        let media, nextPageToken;
+        let media, newNextPageToken;
         if (selectedCategoryId) {
-          ({ media, nextPageToken } = await listCategoryMedia({
-            provider,
-            selectedCategoryId,
-            pageToken,
-          }));
+          ({ media, nextPageToken: newNextPageToken } = await listCategoryMedia(
+            {
+              provider,
+              selectedCategoryId,
+              pageToken,
+            }
+          ));
         } else {
-          ({ media, nextPageToken } = await listMedia({
+          ({ media, nextPageToken: newNextPageToken } = await listMedia({
             provider,
             searchTerm,
             pageToken,
           }));
         }
-        fetchMediaSuccess({ provider, media, pageToken, nextPageToken });
+        fetchMediaSuccess({
+          provider,
+          media,
+          pageToken,
+          nextPageToken: newNextPageToken,
+        });
       } catch {
         fetchMediaError({ provider, pageToken });
         showSnackbar({ message: getFetchMediaErrorMessage(provider) });
       }
     }
 
-    if (provider === selectedProvider) {
+    // If we switched provider tab, and we already had loaded a page there, we
+    // don't load media again.
+    const somethingChanged =
+      previousProps &&
+      (pageToken != previousProps.pageToken ||
+        searchTerm != previousProps.searchTerm ||
+        selectedCategoryId != previousProps.selectedCategoryId);
+    const firstFetchOrSomethingChanged =
+      (!pageToken && !nextPageToken) || somethingChanged;
+
+    if (provider === selectedProvider && firstFetchOrSomethingChanged) {
       fetch();
     }
+    // We don't want to depend on previousProps, see https://blog.logrocket.com/how-to-get-previous-props-state-with-react-hooks/
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
Another way to avoid depending on previousProps, you could use the previousProps reference inside of the useEffect block:

```
  const previousPropsRef = useRef();
  useEffect(() => {
    const previousProps = previousPropsRef.current;
    previousPropsRef.current = { pageToken, searchTerm, selectedCategoryId };
    // ...
  }, [
    // The ref never changes
    previousPropsRef
  ]);
```

Reply:
I'll try this out and report back.

Reply:
I don't think this would do what we want. Accessing previousPropsRef.current inside the effect would not give you the previous value of the props as it's updated inside an effect in usePrevious.

Reply:
The usePrevious hook would not be used in the example given above.
previousPropsRef.current is updated in the useFetchMediaEffect's useEffect().